contracts/token/sPrime.sol

```
@@ -0,0 +1,440 @@
+        uint256 amountY
+    ) internal {
+        PairInfo memory pair = pairList[binStep][activeId];
+
+        require(amountX != 0 || amountY != 0, "ZeroAmounts");
+
+        // Get the lbPair address and transfer the tokens to the lbPair.
+        address lbPair = pair.lbPair;
+
+        if (amountX > 0) tokenX.safeTransfer(lbPair, amountX);
+        if (amountY > 0) tokenY.safeTransfer(lbPair, amountY);
+
+        // Mint the liquidity tokens.
+        ILBPair(lbPair).mint(address(this), liquidityConfigs, address(this));
+
+        (uint256 totalBalanceX, uint256 totalBalanceY) = _getBalances(binStep, activeId);
+        uint256 share = pair.totalShare * _getTotalWeight(lbPair, amountX, amountY) / _getTotalWeight(lbPair, totalBalanceX, totalBalanceY);
+
+        _mint(_msgSender(), share);
+
+        pair.totalShare += share;
+        
+    }
+
+    /**
+     * @dev Withdraws tokens from the lbPair and applies the AUM annual fee. This function will also reset the range.
+     * Will never charge for more than a day of AUM fees, even if the sPrime scontract has not been rebalanced for a longer period.
+     * @param binStep The binStep for the pair.
+     * @param activeId The active Id of the pair
+     * @param share The amount of share to withdraw.
+     * @return totalBalanceX The amount of token X withdrawn.
+     * @return totalBalanceY The amount of token Y withdrawn.
+     */
+    function _withdrawAndResetRange(uint16 binStep, uint24 activeId, uint256 share) internal returns(uint256 totalBalanceX, uint256 totalBalanceY) {
+        PairInfo storage pair = pairList[binStep][activeId];
+
+        // Get the range and reset it.
+        (uint24 lowerRange, uint24 upperRange) = (pair.lowerRange, pair.upperRange);
+        if (upperRange > 0) _resetRange(binStep, activeId);
+
+        (totalBalanceX, totalBalanceY) = _withdrawFromLB(pair.lbPair, lowerRange, upperRange, share * _PRECISION / pair.totalShare);
+
+        _burn(_msgSender(), share);
```

Maybe it would be better to `_burn()` before withdrawing just to fail faster if the `share` is > `balanceOf()`?

---

Fixed it.